modules/gateway/peers.go

```diff
@@ -509,84 +483,110 @@ func (g *Gateway) Connect(addr modules.NetAddress) error {
 	return nil
 }
 
-// Disconnect terminates a connection to a peer and removes it from the
-// Gateway's peer list. The peer's address remains in the node list.
-func (g *Gateway) Disconnect(addr modules.NetAddress) error {
-	if err := g.threads.Add(); err != nil {
-		return err
-	}
-	defer g.threads.Done()
-
-	g.mu.RLock()
-	p, exists := g.peers[addr]
-	g.mu.RUnlock()
-	if !exists {
-		return errors.New("not connected to that node")
-	}
-	g.mu.Lock()
-	delete(g.peers, addr)
-	g.mu.Unlock()
-	if err := p.sess.Close(); err != nil {
-		return err
-	}
-
-	g.log.Println("INFO: disconnected from peer", addr)
-	return nil
-}
-
-// threadedPeerManager tries to keep the Gateway well-connected. As long as
+// permanentPeerManager tries to keep the Gateway well-connected. As long as
 // the Gateway is not well-connected, it tries to connect to random nodes.
-func (g *Gateway) threadedPeerManager() {
-	if g.threads.Add() != nil {
-		return
-	}
-	defer g.threads.Done()
+func (g *Gateway) permanentPeerManager(closedChan chan struct{}) {
+	// Send a signal upon shutdown.
+	defer close(closedChan)
 
 	for {
-		// If we are well-connected, sleep in increments of five minutes until
-		// we are no longer well-connected.
-		g.mu.RLock()
+		// Determine the number of outbound peers.
 		numOutboundPeers := 0
+		g.mu.RLock()
 		for _, p := range g.peers {
 			if !p.Inbound {
 				numOutboundPeers++
 			}
 		}
-		addr, err := g.randomNode()
 		g.mu.RUnlock()
-		if numOutboundPeers >= modules.WellConnectedThreshold {
+		// If the gateway is well connected, sleep for a while and then try
+		// again.
+		if numOutboundPeers >= wellConnectedThreshold {
 			select {
-			case <-time.After(5 * time.Minute):
+			case <-time.After(wellConnectedDelay):
 			case <-g.threads.StopChan():
+				// Interrupt the thread if the shutdown signal is issued.
 				return
 			}
 			continue
 		}
 
-		// Try to connect to a random node. Instead of blocking on Connect, we
-		// spawn a goroutine and sleep for five seconds. This allows us to
-		// continue making connections if the node is unresponsive.
-		if err == nil {
-			go func() {
-				if g.threads.Add() != nil {
-					return
-				}
-				defer g.threads.Done()
-
-				connectErr := g.Connect(addr)
-				if connectErr != nil {
-					g.log.Debugln("WARN: automatic connect failed:", connectErr)
-				}
-			}()
+		// Fetch a random peer.
+		g.mu.RLock()
+		addr, err := g.randomNode()
+		g.mu.RUnlock()
+		// If there was an error, log the error and then wait a while before
+		// trying again.
+		if err != nil {
+			select {
+			case <-time.After(noPeersDelay):
+			case <-g.threads.StopChan():
+				// Interrupt the thread if the shutdown signal is issued.
+				return
+			}
```

Comment:
this is a bit verbose, but the alternative:
```go
if interrupted := sleepLightly(noPeersDelay, g.threads.StopChan()); interrupted {
    return
}
```
...feels too clever and obscures what's happening behind the scenes.